io-sim/test/Test/Control/Monad/Class/MonadMVar.hs

```diff
@@ -0,0 +1,325 @@
+      ]
+    , testGroup "takeMVar"
+      [ testProperty "fairness (IOSim)" prop_takeMVar_fairness_sim
+      , testCase "blocks on an empty MVar (IOSim)"
+        unit_takeMVar_blocks_on_empty_sim
+      , testCase "blocks on an empty MVar (IO)"
+        unit_takeMVar_blocks_on_empty_io
+      ]
+    , testGroup "tryTakeMVar"
+      [ testCase "does not block on an empty MVar (IOSim)"
+        unit_tryTakeMVar_empty
+      , testCase "does not block on a full MVar (IOSim)"
+        unit_tryTakeMVar_full
+      , testCase "return value on an empty MVar (IOSim)"
+        unit_tryTakeMVar_return_empty_sim
+      , testCase "return value on an full MVar (IOSim)"
+        unit_tryTakeMVar_return_full_sim
+      ]
+    , testGroup "tryPutMVar"
+      [ testCase "does not block on an empty MVar (IOSim)"
+        unit_tryPutMVar_empty
+      , testCase "does not block on a full MVar (IOSim)"
+        unit_tryPutMVar_full
+      , testCase "return value on an empty MVar (IOSim)"
+        unit_tryPutMVar_return_empty_sim
+      , testCase "return value on an full MVar (IOSim)"
+        unit_tryPutMVar_return_full_sim
+      ]
+    , testGroup "isEmptyMVar"
+      [ testCase "empty MVar is empty"    unit_isEmptyMVar_empty_sim
+      , testCase "full MVar is not empty" unit_isEmptyMVar_full_sim
+      ]
+    ]
+
+
+--
+-- putMVar
+--
+
+-- | Check that 'takeMVar' is fair.  This is test is only designed for 'IOSim'
+-- as it relies on its thread scheduling and determinism.
+--
+prop_putMVar_fairness
+  :: forall m.
+     ( MonadAsync m
+     , MonadDelay m
+     , MonadMVar  m
+     )
+  => Int -- ^ number of threads
+  -> m Bool
+prop_putMVar_fairness n = do
+    v  <- newEmptyMVar
+    traverse_ (\a -> async $ do threadDelay 0.01
+                                putMVar v a)
+              [1..n]
+    threadDelay 0.02
+    results <- sequence (replicate n (takeMVar v))
+    return $ results == [1..n]
+
+prop_putMVar_fairness_sim :: Positive (Small Int)
+                          -> Property
+prop_putMVar_fairness_sim (Positive (Small n)) =
+    let trace = runSimTrace (prop_putMVar_fairness n)
+    in counterexample (ppTrace trace)
+     $ case traceResult False trace of
+        Left err -> counterexample (show err) False
+        Right a  -> property a
+
+
+unit_putMVar_blocks_on_full
+  :: ( MonadFork  m
+     , MonadDelay m
+     , MonadMVar  m
+     , MonadMonotonicTime m
+     )
+  => m Bool
+unit_putMVar_blocks_on_full = do
+    start <- getMonotonicTime
+    let delta = 0.01
+    v :: MVar m () <- newMVar ()
+    _ <- forkIO $ threadDelay delta
+               >> takeMVar v
+               $> ()
+    putMVar v ()
+    end <- getMonotonicTime
+    return (end `diffTime` start >= delta)
+
+unit_putMVar_blocks_on_full_sim :: Assertion
+unit_putMVar_blocks_on_full_sim = assertBool "did not block on an full MVar" $
+    runSimOrThrow unit_putMVar_blocks_on_full
+
+unit_putMVar_blocks_on_full_io :: Assertion
+unit_putMVar_blocks_on_full_io =
```

Comment:
Maybe add `prop_` prefix to these?

Reply:
For unitests we've been using `unit_` prefix.